File: neetcode150/linkedList/2-addTwoNumbers.js
// 2. Add Two Numbers
// You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order, and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.
// You may assume the two numbers do not contain any leading zero, except the number 0 itself.

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) { 
    let dummy = new ListNode();
    let curr = dummy;
    let carry = 0;
    while (l1 !== null || l2 !== null || carry === 1) {
        let val1, val2;
        if (l1 !== null) val1 = l1.val;
        else val1 = 0;
        if (l2 !== null) val2 = l2.val;
        else val2 = 0;

        let sum = val1 + val2 + carry;
        carry = Math.floor(sum / 10);
        let remain = sum % 10;
        let newNode = ListNode(remain);
        curr.next = newNode;

        if (l1 !== null) l1 = l1.next;
        if (l2 !== null) l2 = l2.next;
        curr = curr.next;
    }
    return dummy.next;
}
// var addTwoNumbers = function(l1, l2) {
//     let copyL1 = l1;
//     let copyL2 = l2;
//     let dummy = new ListNode();
//     let curr = dummy;
//     let carry = 0;
//     while (copyL1 !== null && copyL2 !== null) {
//         let remaining = 0;
//         let sum = copyL1.val + copyL2.val + carry
//         remaining = sum % 10;
//         let newNode = new ListNode(remaining);
//         if (sum >= 10) {
//             carry = 1;
//         } else {
//             carry = 0;
//         }
//         curr.next = newNode;
//         curr = curr.next;
//         copyL1 = copyL1.next;
//         copyL2 = copyL2.next;
//     }
//     while (copyL1 !== null) {
//         let sum = copyL1.val + carry;
//         let newNode = new ListNode(sum % 10);
//         if (sum >= 10) {
//             carry = 1;
//         } else {
//             carry = 0;
//         }
//         curr.next = newNode;
//         curr = curr.next;
//         copyL1 = copyL1.next;
//     }
//     while (copyL2 !== null) {
//         let sum = copyL2.val + carry;
//         let newNode = new ListNode(sum % 10);
//         if (sum >= 10) {
//             carry = 1;
//         } else {
//             carry = 0;
//         }
//         curr.next = newNode;
//         curr = curr.next;
//         copyL2 = copyL2.next;
//     }
//     if (carry === 1) {
//         curr.next = new ListNode(carry);
//         curr = curr.next;
//     }
//     return dummy.next;
// };

let l1 = [2,4,3], l2 = [5,6,4];
let l3 = [0], l4 = [0];
let l5 = [9,9,9,9,9,9,9], l6 = [9,9,9,9];
console.log(addTwoNumbers(l1, l2));
console.log(addTwoNumbers(l3, l4));
console.log(addTwoNumbers(l5, l6));

File: specified-learning专项训练/linkedList/2-addTwoNumbers.py
# 2. Add Two Numbers
# You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order, and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.
# You may assume the two numbers do not contain any leading zero, except the number 0 itself.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

from typing import Optional
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode
        curr = dummy
        carry = 0
        while l1 != None or l2 != None or carry != 0:
            v1 = l1.val if l1 != None else 0
            v2 = l2.val if l2 != None else 0
            total = v1 + v2 + carry
            reste = total % 10
            carry = total // 10 

            curr.next = ListNode(reste)
            l1 = l1.next if l1 != None else None
            l2 = l2.next if l2 != None else None
            curr = curr.next
        return dummy.next

    # 我的方法
    # def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    #     dummy = ListNode()
    #     tail = dummy
    #     addin = 0
    #     while newHead1 != None and newHead2 != None:
    #         totalVal = newHead1.val + newHead2.val + addin
    #         addin = int(totalVal / 10)
    #         reste = totalVal % 10
    #         tail.next = ListNode(reste)
    #         newHead1 = newHead1.next
    #         newHead2 = newHead2.next
    #         tail = tail.next
    #     while newHead1 != None:
    #         totalVal = newHead1.val + addin
    #         addin = int(totalVal / 10)
    #         reste = totalVal % 10 
    #         tail.next = ListNode(reste)
    #         newHead1 = newHead1.next
    #         tail = tail.next
    #     while newHead2 != None:
    #         totalVal = newHead2.val + addin
    #         addin = int(totalVal / 10)
    #         reste = totalVal % 10 
    #         tail.next = ListNode(reste)
    #         newHead2 = newHead2.next
    #         tail = tail.next
    #     if addin != 0:
    #         tail = ListNode(addin)
    #     return dummy.next